data_model/src/fixed.rs

```diff
@@ -0,0 +1,265 @@
+use std::{convert::TryFrom, mem::size_of};
+
+use fixnum::{
+    ops::{CheckedAdd, CheckedSub, Zero},
+    typenum::U9,
+    ArithmeticError, ConvertError, FixedPoint,
+};
+use iroha_schema::prelude::*;
+use parity_scale_codec::{Decode, Encode, Error, Input, Output};
+use serde::{Deserialize, Serialize};
+use thiserror::Error;
+
+/// Base type for fixed implementation. May be changed in forks.
+/// To change implementation to i128 or other type you will need to change it in Cargo.toml.
+type Base = i64;
+
+/// Signed fixed point amount over 64 bits, 9 decimal places.
+///
+/// MAX = (2 ^ (`BITS_COUNT` - 1) - 1) / 10 ^ PRECISION =
+///     = (2 ^ (64 - 1) - 1) / 1e9 =
+///     = 9223372036.854775807 ~ 9.2e9
+/// `ERROR_MAX` = 0.5 / (10 ^ PRECISION) =
+///           = 0.5 / 1e9 =
+///           = 5e-10
+pub type FixNum = FixedPoint<Base, U9>;
+
+/// An encapsulation of [`Fixed`] in encodable form. [`Fixed`] values
+/// should never become negative.
+#[derive(Clone, Copy, Debug, Serialize, Deserialize, IntoSchema, Eq)]
+pub struct Fixed(FixNum);
+
+impl Fixed {
+    /// Constant, representing zero value
+    pub const ZERO: Fixed = Fixed(FixNum::ZERO);
+
+    #[inline]
+    #[cfg(test)]
+    /// Return the only possible negative [`Fixed`] value. Only used for tests.
+    ///
+    /// # Panics
+    /// Never.
+    pub fn negative_one() -> Self {
+        #[allow(clippy::unwrap_used)]
+        Self("-1".parse().unwrap())
+    }
+
+    #[inline]
+    /// Checks if this instance is zero
+    pub const fn is_zero(self) -> bool {
+        *self.0.as_bits() == Base::ZERO
+    }
+
+    #[inline]
+    fn valid(self) -> Result<Self, FixedPointOperationError> {
+        if self > Self::ZERO || self.is_zero() {
+            Ok(self)
+        } else {
+            Err(FixedPointOperationError::NegativeValue(self.0))
+        }
+    }
+
+    #[inline]
+    /// Checked addition
+    ///
+    /// # Errors
+    /// If either of the operands is negative or if addition overflows.
+    pub fn checked_add(self, rhs: Self) -> Result<Self, FixedPointOperationError> {
+        match self.valid()?.0.cadd(rhs.valid()?.0) {
+            Ok(n) => Ok(Fixed(n)),
+            Err(e) => Err(e.into()),
+        }
+    }
+
+    #[inline]
+    /// Checked subtraction
+    ///
+    /// # Errors
+    /// If either of the operands is negative or if the subtraction overflows.
+    pub fn checked_sub(self, rhs: Self) -> Result<Self, FixedPointOperationError> {
+        match self.valid()?.0.csub(rhs.valid()?.0) {
+            Ok(n) => Fixed(n).valid(),
+            Err(e) => Err(e.into()),
+        }
+    }
+}
+
+/// Custom error type for Fixed point operation errors.
+#[derive(Debug, Clone, Error)]
+pub enum FixedPointOperationError {
+    /// All [`Fixed`] values should be positive.
+    #[error("Negative value {0}")]
+    NegativeValue(FixNum),
+    /// Conversion failed.
+    #[error("Failed to produce fixed point number")]
+    FixNum(#[source] ConvertError),
```

Comment:
could we name this variant more in a way it reflects it's meaning? like `Conversion` or something similar